File: backend/tests/conversations.py
# SPDX-License-Identifier: Apache-2.0
# Copyright The Robinauts Authors

"""Builders for the conversation format, so a test says only what it is about.

Fixed ids and a fixed clock: a test that wants two conversations, a second
run or a particular order says so, and everything else is the same in every
test that reads them.
"""

from __future__ import annotations

import uuid
from datetime import UTC, datetime, timedelta

from robinauts.domain import (
    AgentDefinition,
    Channel,
    Conversation,
    Engine,
    Message,
    MessagePart,
    Provenance,
    Role,
    Run,
    RunState,
    TextPart,
)

CONVERSATION = uuid.UUID("11111111-1111-4111-8111-111111111111")
OTHER_CONVERSATION = uuid.UUID("22222222-2222-4222-8222-222222222222")
OWNER = uuid.UUID("33333333-3333-4333-8333-333333333333")
RUN = uuid.UUID("44444444-4444-4444-8444-444444444444")
AGENT = "assistant"
MODEL = "sonnet"

T0 = datetime(2026, 9, 21, 9, 0, tzinfo=UTC)
"""When every test conversation begins."""


def at(seconds: float) -> datetime:
    """``seconds`` after the start; what orders the messages of a test."""
    return T0 + timedelta(seconds=seconds)


def agent_definition(**changes: object) -> AgentDefinition:
    """The agent every test that is not about agents runs."""
    fields: dict[str, object] = {
        "id": AGENT,
        "title": "Assistant",
        "system_prompt": "Play fair.",
        "model": MODEL,
        "engine": Engine.PYDANTIC_AI,
    }
    fields.update(changes)
    return AgentDefinition(**fields)  # type: ignore[arg-type]


def provenance(**changes: object) -> Provenance:
    fields: dict[str, object] = {
        "agent": AGENT,
        "engine": Engine.PYDANTIC_AI,
        "model": MODEL,
        "run_id": RUN,
    }
    fields.update(changes)
    return Provenance(**fields)  # type: ignore[arg-type]


def _parent_id(parent: Message | uuid.UUID | None) -> uuid.UUID | None:
    return parent.id if isinstance(parent, Message) else parent


def question(
    text: str = "What is a robinaut?",
    *,
    parent: Message | uuid.UUID | None = None,
    seconds: float = 0.0,
    conversation_id: uuid.UUID = CONVERSATION,
    parts: tuple[MessagePart, ...] | None = None,
    **changes: object,
) -> Message:
    """A message from a person."""
    fields: dict[str, object] = {
        "id": uuid.uuid4(),
        "conversation_id": conversation_id,
        "parent_id": _parent_id(parent),
        "role": Role.USER,
        "parts": parts if parts is not None else (TextPart(text),),
        "created_at": at(seconds),
        "channel": Channel.WEB,
    }
    fields.update(changes)
    return Message(**fields)  # type: ignore[arg-type]


def answer(
    parent: Message | uuid.UUID,
    text: str = "Someone who plays fair.",
    *,
    seconds: float = 1.0,
    conversation_id: uuid.UUID = CONVERSATION,
    parts: tuple[MessagePart, ...] | None = None,
    **changes: object,
) -> Message:
    """A message from an agent, with the provenance an answer records."""
    fields: dict[str, object] = {
        "id": uuid.uuid4(),
        "conversation_id": conversation_id,
        "parent_id": _parent_id(parent),
        "role": Role.ASSISTANT,
        "parts": parts if parts is not None else (TextPart(text),),
        "created_at": at(seconds),
        "channel": Channel.WEB,
        "provenance": provenance(),
    }
    fields.update(changes)
    return Message(**fields)  # type: ignore[arg-type]


def conversation(**changes: object) -> Conversation:
    fields: dict[str, object] = {
        "id": CONVERSATION,
        "owner_id": OWNER,
        "agent": AGENT,
        "created_at": at(0),
        "updated_at": at(0),
        "title": "What is a robinaut?",
    }
    fields.update(changes)
    return Conversation(**fields)  # type: ignore[arg-type]


def run(**changes: object) -> Run:
    fields: dict[str, object] = {
        "id": RUN,
        "conversation_id": CONVERSATION,
        "message_id": uuid.uuid4(),
        "agent": AGENT,
        "engine": Engine.PYDANTIC_AI,
        "model": MODEL,
        "state": RunState.RUNNING,
        "created_at": at(0),
        "started_at": at(0),
    }
    fields.update(changes)
    return Run(**fields)  # type: ignore[arg-type]


def ended(state: RunState, **changes: object) -> Run:
    """A run that ended in ``state``, with what that state asks for."""
    fields: dict[str, object] = {"state": state, "finished_at": at(2)}
    if state is RunState.FAILED:
        fields["error"] = "the provider said no"
    fields.update(changes)
    return run(**fields)

File: frontend/src/test/conversations.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright The Robinauts Authors

/**
 * Conversations to test with, in the shapes the API answers in.
 *
 * Shared by the tests of the history, of the panel's list and of the shell's
 * routing, so that one change to the wire is one change here.
 */
import type { components } from "../api/schema";
import type { Conversation, Message } from "../conversation/conversation";

/** What opening a conversation answers with. */
type Opened = components["schemas"]["OpenedConversationResponse"];

/** The run a conversation that is answering is answering with. */
export const RUN = "11111111-2222-4333-8444-555555555555";

/** A conversation id of the shape the router insists on. */
export function id(n: number): string {
  return `00000000-0000-4000-8000-${String(n).padStart(12, "0")}`;
}

export function conversation(
  n: number,
  title = `Conversation ${n}`,
): Conversation {
  return {
    id: id(n),
    title,
    agent: "helper",
    created_at: "2026-09-01T10:00:00Z",
    updated_at: "2026-09-02T10:00:00Z",
    active_leaf_id: null,
  };
}

/** A page of them, newest first, with or without a cursor to the next. */
export function page(
  items: Conversation[],
  nextCursor: string | null = null,
): { items: Conversation[]; next_cursor: string | null } {
  return { items, next_cursor: nextCursor };
}

/** One message of a conversation's tree. */
export function message(
  messageId: string,
  parentId: string | null,
  role: "user" | "assistant",
  text: string,
  provenance: Message["provenance"] = null,
): Message {
  return {
    id: messageId,
    parent_id: parentId,
    role,
    channel: "web",
    created_at: "2026-09-02T09:30:00Z",
    parts: [{ kind: "text", text }],
    provenance,
  };
}

/** What an answer records, on an assistant message. */
export function madeBy(
  agent = "helper",
  engine: "langgraph" | "pydantic-ai" = "langgraph",
): Message["provenance"] {
  return { agent, engine, model: "sonnet", run_id: RUN };
}

/** A conversation as `GET /api/conversations/{id}` answers it. */
export function opened(
  summary: Conversation,
  messages: Message[],
  leafId: string | null,
  rest: Partial<Opened> = {},
): Opened {
  return {
    conversation: summary,
    messages,
    leaf_id: leafId,
    run_id: null,
    resume: null,
    ended_badly: null,
    ...rest,
  };
}
